File: python/semantic_kernel/contents/reasoning_content.py
# Copyright (c) Microsoft. All rights reserved.

"""Reasoning content model."""

from html import unescape
from typing import ClassVar, Literal, TypeVar
from xml.etree.ElementTree import Element  # nosec

from pydantic import Field

from semantic_kernel.contents.const import REASONING_CONTENT_TAG, ContentTypes
from semantic_kernel.contents.kernel_content import KernelContent
from semantic_kernel.exceptions.content_exceptions import ContentInitializationError

_T = TypeVar("_T", bound="ReasoningContent")


class ReasoningContent(KernelContent):
    """Represents reasoning content.

    Exposes a human-readable reasoning ``text``. Any provider-specific fields (for example: ids, encrypted blobs,
    statuses, token info) must be carried in ``metadata`` on the base ``KernelContent``.

    Attributes:
        content_type: Literal identifying this instance as reasoning content.
        tag: XML tag name used when serializing to/from XML.
        text: The reasoning text to surface to callers.

    Methods:
        __str__: Return the reasoning text.
        to_element: Serialize to an XML Element using ``tag`` and ``text``.
        from_element: Deserialize from an XML Element into a ReasoningContent.
        to_dict: Serialize to a dict suitable for message payloads.
    """

    content_type: Literal[ContentTypes.REASONING_CONTENT] = Field(REASONING_CONTENT_TAG, init=False)
    tag: ClassVar[str] = REASONING_CONTENT_TAG
    text: str | None = None

    def __str__(self) -> str:
        """Return the text of the reasoning content."""
        return self.text or ""

    def to_element(self) -> Element:
        """Convert the instance to an XML Element."""
        element = Element(self.tag)
        element.text = self.text
        return element

    @classmethod
    def from_element(cls: type[_T], element: Element) -> _T:
        """Create an instance from an XML Element."""
        if element.tag != cls.tag:
            raise ContentInitializationError(f"Element tag is not {cls.tag}")  # pragma: no cover
        return cls(text=unescape(element.text) if element.text else "")

    def to_dict(self) -> dict[str, str]:
        """Convert the instance to a dictionary suitable for message serialization."""
        return {"type": "reasoning", "text": self.text or ""}

File: nodejs/semantic-kernel/contents/reasoning-content.ts
import { REASONING_CONTENT_TAG } from './const'
import { KernelContent } from './kernel-content'

/**
 * Represents reasoning content.
 *
 * Exposes a human-readable reasoning text. Any provider-specific fields (for example: ids, encrypted blobs,
 * statuses, token info) must be carried in metadata on the base KernelContent.
 */
export class ReasoningContent extends KernelContent {
  readonly contentType: string = REASONING_CONTENT_TAG
  readonly tag: string = REASONING_CONTENT_TAG
  text?: string

  constructor(
    options: {
      text?: string
      innerContent?: any
      aiModelId?: string
      metadata?: Record<string, any>
    } = {}
  ) {
    const { text, innerContent, aiModelId, metadata } = options
    super({ innerContent, aiModelId, metadata })
    this.text = text
  }

  /**
   * Return the text of the reasoning content.
   */
  toString(): string {
    return this.text || ''
  }

  /**
   * Convert the instance to an XML Element.
   */
  toElement(): { tag: string; attributes: Record<string, string>; text?: string } {
    return {
      tag: this.tag,
      attributes: {},
      text: this.text,
    }
  }

  /**
   * Create an instance from an XML Element.
   */
  static fromElement(element: { tag: string; attributes: Record<string, string>; text?: string }): ReasoningContent {
    if (element.tag !== REASONING_CONTENT_TAG) {
      throw new Error(`Element tag is not ${REASONING_CONTENT_TAG}`)
    }

    // Unescape HTML entities in the text
    const text = element.text ? this.unescapeHtml(element.text) : ''

    return new ReasoningContent({ text })
  }

  /**
   * Convert the instance to a dictionary suitable for message serialization.
   */
  toDict(): Record<string, any> {
    return {
      type: 'reasoning',
      text: this.text || '',
    }
  }

  /**
   * Unescape HTML entities.
   */
  private static unescapeHtml(text: string): string {
    const htmlEntities: Record<string, string> = {
      '&amp;': '&',
      '&lt;': '<',
      '&gt;': '>',
      '&quot;': '"',
      '&#39;': "'",
      '&nbsp;': ' ',
    }

    return text.replace(/&[^;]+;/g, (entity) => htmlEntities[entity] || entity)
  }
}
